huaweicloud/services/hss/resource_huaweicloud_hss_host_protection.go

```diff
@@ -0,0 +1,352 @@
+	if agentStatus != hostAgentStatusOnline {
+		return fmt.Errorf("the host anget status for HSS protection must be: online,"+
+			" but the current host (%s) agent status is: %s ", hostId, agentStatus)
+	}
+
+	return nil
+}
+
+func switchHostsProtectStatus(client *hssv5.HssClient, region, epsId, hostId string, d *schema.ResourceData) error {
+	var (
+		version      = d.Get("version").(string)
+		chargingMode = d.Get("charging_mode").(string)
+		quotaId      = d.Get("quota_id").(string)
+	)
+
+	if chargingMode == chargingModePrePaid {
+		chargingMode = chargingModePacketCycle
+	}
+	if chargingMode == chargingModePostPaid {
+		chargingMode = chargingModeOnDemand
+	}
+
+	switchOpts := hssv5model.SwitchHostsProtectStatusRequest{
+		Region:              region,
+		EnterpriseProjectId: utils.StringIgnoreEmpty(epsId),
+		Body: &hssv5model.SwitchHostsProtectStatusRequestInfo{
+			Version:      version,
+			ChargingMode: utils.String(chargingMode),
+			ResourceId:   utils.StringIgnoreEmpty(quotaId),
+			HostIdList:   []string{hostId},
+		},
+	}
+
+	_, err := client.SwitchHostsProtectStatus(&switchOpts)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceHostProtectionCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg    = meta.(*config.Config)
+		region = cfg.GetRegion(d)
+		epsId  = common.GetEnterpriseProjectID(d, cfg)
+		hostId = d.Get("host_id").(string)
+	)
+
+	client, err := cfg.HcHssV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating HSS v5 client: %s", err)
+	}
+
+	checkHostAvailableErr := checkHostAvailable(client, region, epsId, hostId)
+	if checkHostAvailableErr != nil {
+		return diag.FromErr(checkHostAvailableErr)
+	}
+
+	err = switchHostsProtectStatus(client, region, epsId, hostId, d)
+	if err != nil {
+		return diag.Errorf("error opening HSS host (%s) protection: %s", hostId, err)
+	}
+
+	d.SetId(hostId)
+
+	return resourceHostProtectionRead(ctx, d, meta)
+}
+
+func resourceHostProtectionRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg    = meta.(*config.Config)
+		region = cfg.GetRegion(d)
+		id     = d.Id()
+		epsId  = common.GetEnterpriseProjectID(d, cfg)
+	)
+
+	client, err := cfg.HcHssV5Client(region)
+	if err != nil {
+		return diag.Errorf("error creating HSS v5 client: %s", err)
+	}
+
+	// If the enterprise project ID is not set during query, query all enterprise projects.
+	if epsId == "" {
+		epsId = queryAllEpsValue
+	}
+	listHostOpts := hssv5model.ListHostStatusRequest{
+		Region:              &region,
+		EnterpriseProjectId: utils.String(epsId),
+		HostId:              utils.String(id),
+	}
+
+	resp, err := client.ListHostStatus(&listHostOpts)
+	if err != nil {
+		return diag.Errorf("error querying HSS hosts: %s", err)
+	}
+
+	if resp == nil || resp.DataList == nil {
+		return diag.Errorf("the host (%s) for HSS host protection does not exist", id)
+	}
+
+	hostList := *resp.DataList
```

Comment:
Please consider the scenario where the hostList length is 0.

Reply:
fixed.